### PR TITLE
Replace git shallow clone with shutil.copytree

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This is a rebase of https://github.com/ansible/awx/pull/7367 (but actually re-applied by hand)

As of opening this:
 - I have some tests running, and don't want to take further action until looking at those
 - I would like to break off a new method `build_project_dir`, for a couple of reasons
   - so we don't erroneously mix `pre_run_hook` error handling into the project error handling
   - due to this change, we need to more clearly demarcate where the `project` folder needs to be created (copytree requires that it does not yet exist)
   - this is more descriptive and move more code out of the catch-all `*_hook` methods

Because of the heavy tie-ins to this change, I believe it's a good time to do that minor refactor after the basic functionality is validated.

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - API

##### ADDITIONAL INFORMATION
I talked before about removing the GitPython library. For the moment, I am considering that too complex and risky. We have two very narrow uses for that library that remain. The objective here is just to fix the git-related bugs.

### CODE REVIEW

File: awx/main/tasks/jobs.py
Comment (AlanCoding) [inline awx/main/tasks/jobs.py]: I'm still not happy with this one part. It's superficial and doesn't really work.

We need to obtain the lock before we check the revision, and we need to keep holding it until after the folder copy is done.

```python
git_repo = git.Repo(project_path)
if job_revision == git_repo.head.commit.hexsha:
  ...
```

Otherwise there's not really any point in this. I didn't really want to make a context manager out of this, but maybe that's what will happen.

**Comment by AlanCoding** [comment]
@fosterseth after looking into the other option, I believe PR is the right path forward. I will wait for your work to land and then rebase this.

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [inline awx/main/tasks/jobs.py:1061]: Do you not need the "if manual" check here like on line 1564? If we do end up using that in more than one place could we just pass in the project type to sync_and_copy and let that method return if the type is manual?

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [inline awx/main/tasks/jobs.py:752]: Do you not need the else clause here like you did on line 716?

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [inline awx/main/tasks/jobs.py]: Why do a sub on 'project_update' vs just having it in the string?

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [inline awx/main/tasks/jobs.py:393]: Do you want a max_wait_time counter in here to prevent an infinite loop?

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [inline awx/main/tasks/jobs.py:384]: Should we do this logic inside get_lock_file method?

Comment (AlanCoding) [reply awx/main/tasks/jobs.py:384]: This isn't really code I'm trying to change, I'm just moving it.

I think a problem with `get_lock_file` raising an exception is that it would preclude the approach here, where it does a `.save()` as a last-ditch fix if it initially returns `None`. On the other hand, I would not want a method named `get_lock_file` to cause a save as a side-effect. So that's my 2-second defense of the structure here, although I'm committed to defending it, I don't have a complete idea of how to restructure it.

File: awx/main/tasks/jobs.py
Comment (AlanCoding) [reply awx/main/tasks/jobs.py:393]: But what would we set it to? From the logger down below, we know it's common in real environments to wait for a long time. We've seen 2 hours of waiting happen. This isn't great, but it's going to be hard to avoid with slow and problematic project and Ansible content.

File: awx/main/tasks/jobs.py
Comment (AlanCoding) [reply awx/main/tasks/jobs.py]: ha, the joys of old code. I'll push a commit to modernize.

File: awx/main/tasks/jobs.py
Comment (AlanCoding) [reply awx/main/tasks/jobs.py:752]: Unfortunately no, this is why I filed https://github.com/ansible/awx/issues/12413

There are a couple of model changes that would make a lot of sense to do for consistency's sake, but we need corresponding UI changes for those. This PR exists to solve the problem of nested submodules in projects, and it got very big so I'm trying to punt on further changes.

We _should_ add the `scm_revision` field _to inventory updates_ and there will then be no `if` here, because the same will work for either a `Job` or `InventoryUpdate`

File: awx/main/tasks/jobs.py
Comment (AlanCoding) [reply awx/main/tasks/jobs.py:1061]: You bring up an interesting point, and it looks like the comment on ln 1566 about manual projects is misleading. The method `sync_and_copy` should already handle manual projects correctly. The super method `build_project_dir` doesn't copy anything, it just creates an empty project directory (for purposes of mounting, workdir, and such).

**Comment by AlanCoding** [comment]
After this merges, we should revisit https://github.com/ansible/awx/issues/1823, as it may become trivial.

**Comment by gundalow** [comment]
@AlanCoding What's needed to get this reviewed & merged

File: awx/main/tasks/jobs.py
Comment (john-westcott-iv) [reply awx/main/tasks/jobs.py:393]: What about a setting like AWX_MAX_PROJECT_LOCK_WAIT_TIME_IN_SECONDS, set to something like 6h as a a starting point?

**Comment by gundalow** [comment]
> @fosterseth after looking into the other option, I believe PR is the right path forward. I will wait for your work to land and then rebase this.

Believe this is referring to https://github.com/ansible/awx/pull/12366 which has been merged.

**Comment by gundalow** [comment]
Seth has kindly offered to review this so we can get it merged.

**Comment by AlanCoding** [comment]
I've been working with @jay-steurer to develop complete integration testing of nested git submodules, which we've never had before. We currently have testing of a few permutations of submodules, but these are all one layer deep. Nested submodules requires a different approach, because you can't hack it together with test-playbooks, you need another github repo, or a different approach.

**Review by fosterseth — APPROVED** [review]

File: awx/main/tasks/jobs.py
Comment (shanemcd) [inline awx/main/tasks/jobs.py:1306]: @branic had some valid feedback that it'd be good to catch errors that happen here and make sure they are surfaced clearly to the user.

**Review by shanemcd — CHANGES_REQUESTED** [review]

File: awx/main/tasks/jobs.py
Comment (branic) [reply awx/main/tasks/jobs.py:1306]: There are a couple of other places copytree is used (lines 1324 and 1337) that should have errors caught as well.

Comment (gundalow) [reply awx/main/tasks/jobs.py:1306]: Is that the last thing before we can get this merged?
Do we require further reviews or testing?

Comment (AlanCoding) [reply awx/main/tasks/jobs.py:1306]: This is run inside of the main try-except block in the `run` method (pre_run_hook previously, another name here). Errors there are saved directly to the `result_traceback` of the job. This shows in the output tab.

If their run involved a project sync, then the traceback may be visible on the project sync (not the job), and this is a common cause of user confusion, but it's still in the UI, and broader UX issues I would like to keep separate from this.

**Review by shanemcd — APPROVED** [review]

**Review by kdelee — APPROVED** [review]

**Pull request merged**
